File: src/application/commands/quality_gate_handler.py
"""Application command handler for quality gate execution.

QualityGateHandler orchestrates running quality gates (lint, types, tests,
fitness) via a GateRunnerProtocol adapter and returns a QualityReport.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from src.domain.models.quality_gate import QualityGate, QualityReport

if TYPE_CHECKING:
    from src.application.ports.quality_gate_port import GateRunnerProtocol
    from src.domain.models.quality_gate import GateResult


class QualityGateHandler:
    """Orchestrates quality gate execution.

    Runs one or more quality gates via a GateRunnerProtocol adapter,
    collecting results into a QualityReport. Continues running
    remaining gates even when earlier gates fail.
    """

    _ALL_GATES: tuple[QualityGate, ...] = (
        QualityGate.LINT,
        QualityGate.TYPES,
        QualityGate.TESTS,
        QualityGate.FITNESS,
    )

    def __init__(self, runner: GateRunnerProtocol) -> None:
        """Initialize with a gate runner adapter.

        Args:
            runner: Infrastructure adapter that executes individual gates.
        """
        self._runner = runner

    def check(
        self,
        gates: tuple[QualityGate, ...] | None = None,
    ) -> QualityReport:
        """Run quality gates and return an aggregated report.

        Args:
            gates: Specific gates to run. None runs all four gates.

        Returns:
            QualityReport with results for each requested gate.
        """
        gates_to_run = gates if gates is not None else self._ALL_GATES
        results: list[GateResult] = []

        for gate in gates_to_run:
            result = self._runner.run(gate)
            results.append(result)

        return QualityReport(results=tuple(results))

File: internal/fitness/application/quality_gate_handler.go
package application

import (
	"context"
	"fmt"

	vo "github.com/alto-cli/alto/internal/shared/domain/valueobjects"
)

// allGates is the default set of quality gates to run.
var allGates = []vo.QualityGate{
	vo.QualityGateLint,
	vo.QualityGateTypes,
	vo.QualityGateTests,
	vo.QualityGateFitness,
}

// QualityGateHandler orchestrates quality gate execution.
// Runs one or more quality gates via a GateRunner adapter,
// collecting results into a QualityReport. Continues running
// remaining gates even when earlier gates fail.
type QualityGateHandler struct {
	runner GateRunner
}

// NewQualityGateHandler creates a new QualityGateHandler with an injected GateRunner.
func NewQualityGateHandler(runner GateRunner) *QualityGateHandler {
	return &QualityGateHandler{runner: runner}
}

// Check runs quality gates and returns an aggregated report.
// If gates is nil, all four gates are run.
func (h *QualityGateHandler) Check(
	ctx context.Context,
	gates []vo.QualityGate,
) (vo.QualityReport, error) {
	gatesToRun := gates
	if gatesToRun == nil {
		gatesToRun = allGates
	}

	var results []vo.GateResult
	for _, gate := range gatesToRun {
		result, err := h.runner.Run(ctx, gate)
		if err != nil {
			return vo.QualityReport{}, fmt.Errorf("run gate %s: %w", gate, err)
		}
		results = append(results, result)
	}

	return vo.NewQualityReport(results), nil
}
